### PR TITLE
Add mask loss

### PR DESCRIPTION
**Pull request by PhysSong**
### Changes
Adds the mask loss calculation function `mask_loss`, which accepts the mask predictor logit, ground truth label, and optional permutation argument. The function returns both the loss for given mask prediction(which is from previous frame) and the best match(permutation) which will be used in the next frame.
The function is combination of `get_batch_order` and `get_mask_loss`. Actually, I misunderstood the logic for this part in the first time, so I merged two functions. Later I noticed I was wrong, but I decided to not split the function.

`nn.CrossEntropyLoss` was used on logits before softmax instead of NLL on classification results, as in the authors' code, but I can change it to use `clf` and `nn.NLLLoss` if you want.
### TODO
An hyperparameter for multiplying two different losses before adding them should be added, as in the authors' implementation.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -24,6 +26,47 @@ class Config:
...
+    """
+    def loss_for_permutation(
+        logit: th.Tensor,
+        mask_gt: th.Tensor,
+        permutation: Tuple[int, ...]
+    ) -> th.Tensor:
+        # CrossEntropyLoss on logits is equivalent to NLL on softmax result
+        logit_permuted = th.stack(
+            [logit[b:b+1, -1]] + [logit[b:b+1, i] for i in permutation], dim=1)
+        return nn.CrossEntropyLoss()(logit_permuted, mask_gt[b:b + 1])
```
Comment (yycho0108) [inline train.py]: you can probably use `nn.functional` to avoid creating an awkward class instance.

File: train.py [hunk 2]
```diff
@@ -24,6 +26,47 @@ class Config:
...
+    B, K, S1, S2, S3 = logit.size()
+    best_mask_order = []
+    for b in range(B):
+        best_loss, best_perm = None, None
+        input_permutation = mask_order[b] if mask_order is not None else None
+        for permutation in itertools.permutations(range(K - 1)):
+            cur_loss = loss_for_permutation(logit, mask_gt, permutation)
+            # If mask_order is given, use that
+            if permutation == input_permutation:
+                loss += cur_loss
```
Comment (yycho0108) [inline train.py:65]: Is this correct? `loss` aggregates both `cur_loss` and `best_loss`?

File: train.py [hunk 3]
```diff
@@ -48,15 +91,20 @@ def main():
...
                 prv_state = outputs['state'].detach()
-                loss = motion_loss(outputs['motion'],
-                                   inputs['scene_flow_3d'].to(device))
+                loss_motion = motion_loss(outputs['motion'],
+                                          inputs['scene_flow_3d'].to(device))
+                loss_mask, mask_order = mask_loss(
+                    outputs['logit'],
+                    inputs['mask_3d'].to(device),
+                    mask_order)
+                loss = loss_motion + loss_mask # TODO weight two losses before add
```
Comment (yycho0108) [inline train.py]: I think the authors use `1.0 * loss_motion + 5.0 * loss_mask` by default.
Although it's probably better to expose it as a hyperparameter rather than a hardcoded value.

File: train.py [hunk 4]
```diff
@@ -24,6 +26,47 @@ class Config:
...
+        logit_permuted = th.stack(
+            [logit[b:b+1, -1]] + [logit[b:b+1, i] for i in permutation], dim=1)
+        return nn.CrossEntropyLoss()(logit_permuted, mask_gt[b:b + 1])
+    loss = 0
+    B, K, S1, S2, S3 = logit.size()
+    best_mask_order = []
+    for b in range(B):
+        best_loss, best_perm = None, None
+        input_permutation = mask_order[b] if mask_order is not None else None
+        for permutation in itertools.permutations(range(K - 1)):
```
Comment (yycho0108) [inline train.py:61]: It's concerning since this operation might be super slow. (`(K-1)!` iterations)
Is this how the authors implemented this matching part?

File: train.py [hunk 5]
```diff
@@ -48,15 +91,20 @@ def main():
         for data in tqdm(loader, leave=False, desc='batch'):
             # Formatted as TxBx[...]
             prv_state = None
+            mask_order = None
```
Comment (yycho0108) [inline train.py:101]: Does `mask_order` get used anywhere?

**Review by yycho0108 — APPROVED** [review]
Overall LGTM, except a few comments.

File: train.py (hunk 2 above)
Comment (PhysSong) [reply train.py:65]: If `input_permutation` is given, `cur_loss` is aggregated. Otherwise, `best_loss` is aggregated.

File: train.py (hunk 1 above)
Comment (PhysSong) [reply train.py]: Good point, I will consider changing.

File: train.py (hunk 4 above)
Comment (PhysSong) [reply train.py:61]: They tries all `(K-1)!` iterations. Since K is small, that doesn't consume much time.

File: train.py (hunk 5 above)
Comment (PhysSong) [reply train.py:101]: As I mentioned before, `mask_loss` gets the best order from previous frame(`None` for the first frame) and returns the best order for the current frame, which will be used in the next frame.

File: train.py (hunk 3 above)
Comment (PhysSong) [reply train.py]: Then I will add two constants `1.0` and `5.0` for now. It will be not that hard to make this configurable after the change.

File: train.py (hunk 5 above)
Comment (yycho0108) [reply train.py:101]: ok.

File: train.py (hunk 4 above)
Comment (yycho0108) [reply train.py:61]: ok

**Comment by PhysSong** [comment]
@yycho0108 I addressed your reviews. Please check if everything is okay now.

**Comment by yycho0108** [comment]
Looks good to me.
